File: backend/app/services/llm_service.py
# llm_service.py - Updated with Current Groq Models
import json
import httpx
import logging
from typing import Dict, Any, Optional
from langchain_groq import ChatGroq
from langchain.schema import HumanMessage, SystemMessage
from langchain.prompts import PromptTemplate
from langchain.chains import LLMChain
from app.config import settings

logger = logging.getLogger(__name__)

class LLMService:
    def __init__(self):
        self.groq_api_key = settings.GROQ_API_KEY
        
        if not self.groq_api_key:
            logger.warning("GROQ_API_KEY not found in config settings")
            self.groq_llm = None
        else:
            # Initialize Groq with LangChain
            try:
                self.groq_llm = ChatGroq(
                    api_key=self.groq_api_key,
                    model="llama-3.1-8b-instant",  # ✅ UPDATED: New current model
                    temperature=0.1,
                    max_tokens=2000,
                    timeout=30
                )
                logger.info("Groq LLM initialized successfully with llama-3.1-8b-instant")
            except Exception as e:
                logger.error(f"Failed to initialize Groq LLM: {str(e)}")
                self.groq_llm = None

    async def call_groq_with_langchain(self, prompt: str, system_message: Optional[str] = None) -> Dict[str, Any]:
        """Call Groq LLM using LangChain"""
        if not self.groq_llm:
            raise ValueError("Groq LLM is not configured or initialized")
        
        messages = []
        if system_message:
            messages.append(SystemMessage(content=system_message))
        messages.append(HumanMessage(content=prompt))
        
        try:
            response = await self.groq_llm.ainvoke(messages)
            try:
                return json.loads(response.content)
            except json.JSONDecodeError:
                return {"result": response.content}
        except Exception as e:
            logger.error(f"Groq LangChain call failed: {str(e)}")
            raise Exception(f"LLM call failed: {str(e)}")

    async def call_groq_direct(self, prompt: str, model: str = "llama-3.1-8b-instant") -> Dict[str, Any]:
        """Call Groq Model via direct API request"""
        if not self.groq_api_key:
            raise ValueError("Groq API key not configured in settings")
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.post(
                    "https://api.groq.com/openai/v1/chat/completions",
                    headers={
                        "Authorization": f"Bearer {self.groq_api_key}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "model": model,  # ✅ UPDATED: Now uses current model by default
                        "messages": [{"role": "user", "content": prompt}],
                        "temperature": 0.1,
                        "max_tokens": 2000
                    },
                    timeout=30.0
                )
                
                if response.status_code != 200:
                    raise Exception(f"Groq API error: {response.status_code} - {response.text}")
                
                result = response.json()
                content = result["choices"][0]["message"]["content"]
                
                try:
                    return json.loads(content)
                except json.JSONDecodeError:
                    return {"result": content}
                    
            except Exception as e:
                logger.error(f"Direct Groq API call failed: {str(e)}")
                raise

    async def call_groq(self, prompt: str, system_message: Optional[str] = None) -> Dict[str, Any]:
        """
        Main Groq calling method with fallback logic
        This is the method that analysis.py is calling
        """
        try:
            # Try LangChain first
            logger.info("Attempting Groq call via LangChain with llama-3.1-8b-instant")
            return await self.call_groq_with_langchain(prompt, system_message)
        except Exception as langchain_error:
            logger.warning(f"Groq LangChain failed: {str(langchain_error)}")
            try:
                # Fallback to direct API
                logger.info("Attempting Groq call via direct API with llama-3.1-8b-instant")
                return await self.call_groq_direct(prompt, "llama-3.1-8b-instant")
            except Exception as direct_error:
                logger.error(f"Direct Groq API also failed: {str(direct_error)}")
                # Return error response instead of raising exception
                return {
                    "error": f"All Groq methods failed: {str(direct_error)}",
                    "analysis": f"Failed to analyze document: {str(direct_error)}",
                    "type": "error"
                }

    async def call_llm_with_fallback(self, prompt: str, system_message: Optional[str] = None) -> Dict[str, Any]:
        """Try Groq with LangChain, fallback to direct Groq API request."""
        try:
            return await self.call_groq_with_langchain(prompt, system_message)
        except Exception as groq_error:
            logger.warning(f"Groq LangChain failed, trying direct API: {str(groq_error)}")
            try:
                return await self.call_groq_direct(prompt, "llama-3.1-8b-instant")
            except Exception as direct_error:
                logger.error(f"Direct Groq API also failed: {str(direct_error)}")
                return {"error": f"All Groq calls failed: {str(direct_error)}"}

    def create_prompt_template(self, template_str: str) -> PromptTemplate:
        """Create a LangChain prompt template."""
        return PromptTemplate.from_template(template_str)

    async def test_connection(self) -> Dict[str, Any]:
        """Test Groq LLM connection and return status."""
        test_prompt = "Reply with 'Connection successful' in JSON format: {\"status\": \"Connection successful\"}"
        
        results = {
            "groq_langchain": "not_tested",
            "groq_direct": "not_tested"
        }
        
        # Test LangChain
        if self.groq_llm:
            try:
                res = await self.call_groq_with_langchain(test_prompt)
                results["groq_langchain"] = "success"
            except Exception as e:
                results["groq_langchain"] = f"failed: {str(e)}"
        
        # Test direct
        if self.groq_api_key:
            try:
                res = await self.call_groq_direct(test_prompt, "llama-3.1-8b-instant")
                results["groq_direct"] = "success"
            except Exception as e:
                results["groq_direct"] = f"failed: {str(e)}"
        
        return {
            "status": "completed",
            "test_results": results,
            "primary_service": "groq",
            "current_model": "llama-3.1-8b-instant"
        }

    async def health_check_groq(self) -> Dict[str, Any]:
        """Test the main call_groq method"""
        try:
            test_result = await self.call_groq("Test message: respond with JSON {\"status\": \"ok\"}")
            return {
                "status": "healthy",
                "method": "call_groq",
                "model": "llama-3.1-8b-instant",
                "response": test_result,
                "timestamp": "2025-09-01T19:22:00"
            }
        except Exception as e:
            return {
                "status": "failed",
                "method": "call_groq", 
                "model": "llama-3.1-8b-instant",
                "error": str(e),
                "timestamp": "2025-09-01T19:22:00"
            }

    # ✅ ADDED: Method to get supported models
    def get_supported_models(self) -> Dict[str, Any]:
        """Get list of current supported Groq models"""
        return {
            "current_models": [
                "llama-3.1-8b-instant",      # Current fast model
                "llama-3.3-70b-versatile",   # Current large model
                "gemma2-9b-it",              # Alternative instruction model
                "mixtral-8x7b-32768",        # Mixtral model
            ],
            "deprecated_models": [
                "llama3-8b-8192",            # Deprecated 8/30/25
                "llama3-70b-8192",           # Deprecated 8/30/25
            ],
            "default_model": "llama-3.1-8b-instant",
            "updated_at": "2025-09-01"
        }

# Global LLM service instance
llm_service = LLMService()

File: backend_node/services/llm_service.js
const axios = require('axios');
const logger = require('../utils/logger');
const config = require('../config');

class EnhancedLLMService {
    constructor() {
        this.groqApiKey = config.GROQ_API_KEY;
        
        if (!this.groqApiKey) {
            logger.warn("GROQ_API_KEY not found in config settings");
            this.groqLlm = null;
        } else {
            try {
                this.modelConfig = {
                    model: "llama-3.3-70b-versatile",
                    temperature: 0.1,
                    max_tokens: 4000,
                    timeout: 90000
                };
                
                this.fallbackConfig = {
                    model: "llama-3.1-8b-instant",
                    temperature: 0.2,
                    max_tokens: 3000,
                    timeout: 45000
                };
                
                logger.info("✅ Enhanced Groq LLM initialized with llama-3.3-70b-versatile");
                this.groqLlm = true;
                
            } catch (error) {
                logger.error(`Failed to initialize Enhanced Groq LLM: ${error.message}`);
                this.groqLlm = null;
            }
        }
    }

    // ✅ CRITICAL FIX: Enhanced JSON parsing with multiple fallback strategies
    safeJsonParse(text) {
        if (!text || typeof text !== 'string') {
            return null;
        }

        // Strategy 1: Direct JSON parse
        try {
            const parsed = JSON.parse(text.trim());
            logger.legalSuccess("✅ Direct JSON parsing successful");
            return parsed;
        } catch (directError) {
            logger.legalWarning("⚠️ Direct JSON parsing failed, trying extraction strategies");
        }

        // Strategy 2: Extract JSON block with code fence
        try {
            const codeBlockMatch = text.match(/``````/i);
            if (codeBlockMatch) {
                const parsed = JSON.parse(codeBlockMatch[1].trim());
                logger.legalSuccess("✅ Code block JSON extraction successful");
                return parsed;
            }
        } catch (codeBlockError) {
            logger.legalWarning("⚠️ Code block JSON extraction failed");
        }

        // Strategy 3: Extract first complete JSON object
        try {
            const jsonMatch = text.match(/\{[\s\S]*\}/);
            if (jsonMatch) {
                const jsonText = jsonMatch[0];
                const parsed = JSON.parse(jsonText);
                logger.legalSuccess("✅ JSON object extraction successful");
                return parsed;
            }
        } catch (objectError) {
            logger.legalWarning("⚠️ JSON object extraction failed");
        }

        // Strategy 4: Fix common JSON issues and retry
        try {
            let fixedText = text
                // Remove any text before first {
                .substring(text.indexOf('{'))
                // Remove any text after last }
                .substring(0, text.lastIndexOf('}') + 1)
                // Fix common quote issues
                .replace(/'/g, '"')
                // Fix trailing commas
                .replace(/,(\s*[}\]])/g, '$1')
                // Fix unescaped quotes in strings
                .replace(/(?<!\\)"/g, '\\"')
                .replace(/\\"/g, '"');

            const parsed = JSON.parse(fixedText);
            logger.legalSuccess("✅ Fixed JSON parsing successful");
            return parsed;
        } catch (fixError) {
            logger.legalWarning("⚠️ Fixed JSON parsing failed");
        }

        // Strategy 5: Lenient JSON-like parsing
        try {
            // Extract key-value pairs and construct JSON
            const keyValuePairs = [];
            
            // Look for "key": "value" patterns
            const patterns = [
                /"([^"]+)":\s*"([^"]*)"/g,
                /"([^"]+)":\s*(\d+\.?\d*)/g,
                /"([^"]+)":\s*(true|false)/g,
                /"([^"]+)":\s*\[([^\]]*)\]/g
            ];

            let result = {};
            
            for (const pattern of patterns) {
                let match;
                while ((match = pattern.exec(text)) !== null) {
                    const key = match[1];
                    let value = match[2];
                    
                    // Type conversion
                    if (value === 'true') value = true;
                    else if (value === 'false') value = false;
                    else if (!isNaN(value) && value !== '') value = parseFloat(value);
                    else if (match[0].includes('[')) {
                        // Array handling
                        try {
                            value = JSON.parse(`[${value}]`);
                        } catch {
                            value = value.split(',').map(v => v.trim().replace(/['"]/g, ''));
                        }
                    }
                    
                    result[key] = value;
                }
            }

            if (Object.keys(result).length > 0) {
                logger.legalSuccess("✅ Lenient JSON parsing successful");
                return result;
            }
        } catch (lenientError) {
            logger.legalWarning("⚠️ Lenient JSON parsing failed");
        }

        logger.legalError("❌ All JSON parsing strategies failed");
        return null;
    }

    async callGroqEnhanced(prompt, systemMessage = null) {
        if (!this.groqLlm) {
            throw new Error("Enhanced Groq LLM is not configured or initialized");
        }

        if (!systemMessage) {
            systemMessage = `You are a senior legal analyst with expertise in contract review, risk assessment, and legal document analysis.

Your responses must be:
- Accurate and based solely on the provided contract content
- Professional and legally sound
- Structured in the exact JSON format requested
- Detailed enough to be actionable

CRITICAL: You MUST respond with valid JSON only. Do not include any text before or after the JSON object. Do not use markdown code blocks. Return raw JSON only.

Always analyze the specific contract terms and conditions provided to you.`;
        }

        const messages = [
            { role: "system", content: systemMessage },
            { role: "user", content: prompt }
        ];

        try {
            logger.aiInfo(`🤖 Calling enhanced LLM with ${prompt.length} character prompt`);

            const maxRetries = 3;
            
            for (let attempt = 0; attempt < maxRetries; attempt++) {
                try {
                    const response = await this._makeGroqRequest(messages, this.modelConfig);
                    const content = response.trim();
                    
                    logger.aiInfo(`🤖 LLM response received: ${content.length} characters`);

                    // ✅ CRITICAL FIX: Use enhanced JSON parsing
                    const parsed = this.safeJsonParse(content);
                    
                    if (parsed) {
                        logger.legalSuccess("✅ LLM response successfully parsed to JSON");
                        return parsed;
                    } else {
                        logger.legalWarning(`⚠️ Attempt ${attempt + 1}: JSON parsing failed, raw response: ${content.substring(0, 200)}...`);
                        
                        if (attempt === maxRetries - 1) {
                            // Create structured fallback response
                            return {
                                success: false,
                                content: content,
                                error: "JSON parsing failed after all attempts",
                                raw_response: content.substring(0, 500)
                            };
                        }
                    }

                } catch (callError) {
                    logger.legalWarning(`⚠️ LLM call attempt ${attempt + 1} failed: ${callError.message}`);
                    if (attempt === maxRetries - 1) {
                        throw callError;
                    }
                    await new Promise(resolve => setTimeout(resolve, 2000));
                }
            }

        } catch (error) {
            logger.legalError(`❌ Enhanced Groq call failed: ${error.message}`);
            return {
                error: `LLM analysis failed: ${error.message}`,
                success: false,
                fallback_required: true
            };
        }
    }

    async callGroqDirectEnhanced(prompt, systemMessage = null) {
        if (!this.groqApiKey) {
            throw new Error("Groq API key not configured");
        }

        try {
            const messages = [];
            if (systemMessage) {
                messages.push({ role: "system", content: systemMessage });
            }
            messages.push({ role: "user", content: prompt });

            const payload = {
                model: "llama-3.3-70b-versatile",
                messages: messages,
                temperature: 0.1,
                max_tokens: 4000,
                top_p: 0.9,
                frequency_penalty: 0,
                presence_penalty: 0
            };

            const response = await axios.post(
                "https://api.groq.com/openai/v1/chat/completions",
                payload,
                {
                    headers: {
                        "Authorization": `Bearer ${this.groqApiKey}`,
                        "Content-Type": "application/json"
                    },
                    timeout: 90000
                }
            );

            if (response.status !== 200) {
                logger.legalError(`Groq API error: ${response.status} - ${response.data}`);
                throw new Error(`Groq API error: ${response.status}`);
            }

            const content = response.data.choices[0].message.content.trim();
            logger.legalSuccess(`✅ Direct API response received: ${content.length} characters`);

            // ✅ CRITICAL FIX: Use enhanced JSON parsing
            const parsed = this.safeJsonParse(content);
            
            if (parsed) {
                return parsed;
            } else {
                // Return structured response even if JSON parsing fails
                return { 
                    result: content, 
                    success: true, 
                    json_parse_failed: true,
                    raw_content: content.substring(0, 1000)
                };
            }

        } catch (error) {
            logger.legalError(`❌ Enhanced direct API call failed: ${error.message}`);
            throw error;
        }
    }

    async _makeGroqRequest(messages, modelConfig) {
        const payload = {
            model: modelConfig.model,
            messages: messages,
            temperature: modelConfig.temperature,
            max_tokens: modelConfig.max_tokens,
            top_p: 0.9,
            frequency_penalty: 0,
            presence_penalty: 0
        };

        const response = await axios.post(
            "https://api.groq.com/openai/v1/chat/completions",
            payload,
            {
                headers: {
                    "Authorization": `Bearer ${this.groqApiKey}`,
                    "Content-Type": "application/json"
                },
                timeout: modelConfig.timeout
            }
        );

        if (response.status !== 200) {
            throw new Error(`Groq API error: ${response.status}`);
        }

        return response.data.choices[0].message.content;
    }

    async callGroq(prompt, systemMessage = null) {
        try {
            return await this.callGroqEnhanced(prompt, systemMessage);
        } catch (langchainError) {
            logger.legalWarning(`⚠️ Enhanced method failed: ${langchainError.message}`);
            try {
                return await this.callGroqDirectEnhanced(prompt, systemMessage);
            } catch (directError) {
                logger.legalError(`❌ All enhanced methods failed: ${directError.message}`);
                
                return {
                    error: "Analysis failed due to LLM service issues",
                    details: directError.message,
                    success: false,
                    fallback_analysis: {
                        risks: [],
                        summary: "Unable to analyze document due to technical issues",
                        emails: {
                            acceptance: "Analysis service temporarily unavailable",
                            rejection: "Analysis service temporarily unavailable"
                        }
                    }
                };
            }
        }
    }

    async healthCheck() {
        try {
            const testResult = await this.callGroq('Test message: respond with JSON {"status": "ok"}');
            return {
                status: "healthy",
                method: "enhanced_call_groq",
                model: "llama-3.3-70b-versatile",
                response: testResult,
                timestamp: new Date().toISOString()
            };
        } catch (error) {
            return {
                status: "failed",
                method: "enhanced_call_groq", 
                model: "llama-3.3-70b-versatile",
                error: error.message,
                timestamp: new Date().toISOString()
            };
        }
    }

    getSupportedModels() {
        return {
            current_model: "llama-3.3-70b-versatile",
            fallback_model: "llama-3.1-8b-instant",
            features: [
                "Enhanced legal analysis",
                "Better accuracy for contract review",
                "Improved JSON response parsing with multiple fallback strategies",
                "Robust error handling with fallbacks"
            ],
            updated_at: "2025-09-13"
        };
    }
}

const llmService = new EnhancedLLMService();

module.exports = llmService;
